tests/unit/test_utils_misc.py

```diff
@@ -0,0 +1,25 @@
+from typing import Dict, List, Union
+
+from pip._internal.utils.misc import merge_config_settings
+
+
+def test_merge_config_settings() -> None:
+    reqs: Dict[str, Union[str, List[str]]] = {
+        "foo": "bar",
+        "bar": "foo",
+        "foobar": ["bar"],
+        "baz": ["foo"],
+    }
+    cli: Dict[str, Union[str, List[str]]] = {
+        "foo": ["baz"],
+        "bar": "bar",
+        "foobar": ["baz"],
+        "baz": "bar",
+    }
+    expected = {
+        "foo": ["bar", "baz"],
+        "bar": ["foo", "bar"],
+        "foobar": ["bar", "baz"],
+        "baz": ["foo", "bar"],
+    }
```

Comment:
There should also be tests for the cases where one settings is in cli or reqs but not the other, with variants where it is a string or a list.

This test is a good candidate to leverage `pytest.mark.parametrize`:)